### PR TITLE
Synching Microsoft.Identity.Web reusable library with aspnetcore webapp tutorial

### PR DESCRIPTION
**Pull request by jmprieur**
- Updating the token cache providers implementations (See https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/106)

- Updating the productized AadIssuerValidator
- Code Improvements

## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: Microsoft.Identity.Web/ClaimConstants.cs [hunk 1]
```diff
@@ -25,12 +25,15 @@
...
     /// <summary>
-    /// claim keys constants
+    /// Constants for claim types.
     /// </summary>
     public static class ClaimConstants
     {
+        public const string Name = "name";
         public const string ObjectId = "http://schemas.microsoft.com/identity/claims/objectidentifier";
+        public const string Oid = "oid";
+        public const string PreferredUserName = "preferred_username";
```
Comment (jennyf19) [inline Microsoft.Identity.Web/ClaimConstants.cs]: >PreferredUserName [](start = 28, length = 17)

nit: PreferredUsername   -> Username is a compound word, so all lowercase

File: Microsoft.Identity.Web/StartupHelpers.cs [hunk 2]
```diff
@@ -1,15 +1,37 @@
+/************************************************************************************************
```
Comment (jennyf19) [inline Microsoft.Identity.Web/StartupHelpers.cs]: Can you use the simplified license header like we do in MSAL? and include the full version as a file? also, the context is different. ex. We don't have a date.

File: Microsoft.Identity.Web/StartupHelpers.cs [hunk 3]
```diff
@@ -112,6 +134,9 @@ public static IServiceCollection AddMsal(this IServiceCollection services, IEnum
             {
                 // Response type
                 options.ResponseType = OpenIdConnectResponseType.CodeIdToken;
+
+                // This scope is needed to get a refresh token when users sign-in with their Microsoft personal accounts
+                // (it's required by MSAL.NET and automatically provided when users sign-in with work or school accounts)
```
Comment (jennyf19) [inline Microsoft.Identity.Web/StartupHelpers.cs]: >( [](start = 19, length = 1)

nit: i don't think you need ( ) here

File: Microsoft.Identity.Web/WebApiStartupHelpers.cs [hunk 4]
```diff
@@ -39,8 +39,8 @@ public static IServiceCollection AddProtectWebApiWithMicrosoftIdentityPlatformV2
                 options.TokenValidationParameters.ValidAudiences = new string[] { options.Audience, $"api://{options.Audience}" };
 
                 // Instead of using the default validation (validating against a single tenant, as we do in line of business apps),
-                // we inject our own multi-tenant validation logic (which even accepts both V1 and V2 tokens)
-                options.TokenValidationParameters.IssuerValidator = AadIssuerValidator.ForAadInstance(options.Authority).ValidateAadIssuer;
+                // we inject our own multitenant validation logic (which even accepts both V1 and V2 tokens)
```
Comment (jennyf19) [inline Microsoft.Identity.Web/WebApiStartupHelpers.cs]: nit: remove ( ) ....put "which accepts both..."

**Comment by jennyf19** [comment]
> using System.Linq;

license?

---
Refers to: Microsoft.Identity.Web/Client/MsalUiRequiredExceptionFilterAttribute.cs:2 in 1b53466. [](commit_id = 1b53466578d35d5ee61edea935d3e7eee9a16d8c, deletion_comment = False)

File: Microsoft.Identity.Web/Client/MsalUiRequiredExceptionFilterAttribute.cs [hunk 5]
```diff
@@ -27,7 +27,13 @@ public class MsalUiRequiredExceptionFilterAttribute : ExceptionFilterAttribute
...
         public override void OnException(ExceptionContext context)
         {
-            if (context.Exception is MsalUiRequiredException msalUiRequiredException)
+            MsalUiRequiredException msalUiRequiredException = context.Exception as MsalUiRequiredException;
+            if (msalUiRequiredException == null)
+            {
+                msalUiRequiredException = context.Exception?.InnerException as MsalUiRequiredException;
+            }
+
+            if (msalUiRequiredException!=null)
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/MsalUiRequiredExceptionFilterAttribute.cs]: >!= [](start = 39, length = 2)

nit: spacing looks off

File: Microsoft.Identity.Web/Client/TokenAcquisition.cs [hunk 6]
```diff
@@ -255,10 +258,12 @@ public async Task RemoveAccount(RedirectContext context)
                 account = accounts.FirstOrDefault(a => a.Username == user.GetLoginHint());
             }
 
- //           this.AppTokenCacheProvider?.Clear();
-            this.UserTokenCacheProvider?.Clear();
+            if (account!=null)
+            {
+                this.UserTokenCacheProvider?.Clear(account.HomeAccountId.Identifier);
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenAcquisition.cs]: >this. [](start = 16, length = 5)

remove .this

**Comment by jennyf19** [comment]
>     public interface IMSALUserTokenCacheProvider

nit: Use Msal -> from Framework design guidelines p. 41 "Acronyms of three or more characters follow the guidelines of any other word. Only the first letter is capitalized"...also we use Msal in MSAL lib and there is inconsistency in the namespaces, ex. MsalUiRequiredExcept...

---
Refers to: Microsoft.Identity.Web/Client/TokenCacheProviders/IMSALUserTokenCacheProvider.cs:34 in 1b53466. [](commit_id = 1b53466578d35d5ee61edea935d3e7eee9a16d8c, deletion_comment = False)

File: Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs [hunk 7]
```diff
@@ -49,17 +49,12 @@ public class MSALAppMemoryTokenCacheProvider : IMSALAppTokenCacheProvider
         /// </summary>
         internal IMemoryCache memoryCache;
 
-        /// <summary>
-        /// The internal handle to the client's instance of the Cache
-        /// </summary>
-        private ITokenCache ApptokenCache;
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs]: >t [](start = 31, length = 1)

nit: capital T

File: Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs [hunk 8]
```diff
@@ -90,12 +85,9 @@ public void Initialize(ITokenCache tokenCache, HttpContext httpcontext)
         {
             this.AppCacheId = this.AppId + "_AppTokenCache";
 
-            this.ApptokenCache = tokenCache;
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs]: >this. [](start = 12, length = 5)

remove .this

File: Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs [hunk 9]
```diff
@@ -90,12 +85,9 @@ public void Initialize(ITokenCache tokenCache, HttpContext httpcontext)
...
-            this.ApptokenCache.SetBeforeAccess(this.AppTokenCacheBeforeAccessNotification);
-            this.ApptokenCache.SetAfterAccess(this.AppTokenCacheAfterAccessNotification);
-            this.ApptokenCache.SetBeforeWrite(this.AppTokenCacheBeforeWriteNotification);
-
-            this.LoadAppTokenCacheFromMemory();
+            tokenCache.SetBeforeAccess(this.AppTokenCacheBeforeAccessNotification);
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALAppMemoryTokenCacheProvider.cs]: >this. [](start = 39, length = 5)

remove .this

File: Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALPerUserMemoryTokenCacheProvider.cs [hunk 10]
```diff
@@ -143,12 +91,21 @@ public void Clear()
...
+                string cacheKey = args.Account?.HomeAccountId?.Identifier;
+                if (string.IsNullOrEmpty(cacheKey))
+                {
+                    cacheKey = httpContextAccessor.HttpContext.User.GetMsalAccountId();
+                }
+
+                if (string.IsNullOrWhiteSpace(cacheKey))
+                    return;
+
+                // Ideally, methods that load and persist should be thread safe.MemoryCache.Get() is thread safe.
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALPerUserMemoryTokenCacheProvider.cs]: space between safe and MemoryCache

File: Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALPerUserMemoryTokenCacheProvider.cs [hunk 11]
```diff
@@ -159,7 +116,17 @@ private void UserTokenCacheAfterAccessNotification(TokenCacheNotificationArgs ar
         /// <param name="args">Contains parameters used by the MSAL call accessing the cache.</param>
         private void UserTokenCacheBeforeAccessNotification(TokenCacheNotificationArgs args)
         {
-            this.LoadUserTokenCacheFromMemory();
+            string cacheKey = args.Account?.HomeAccountId?.Identifier;
+            if (string.IsNullOrEmpty(cacheKey))
+            {
+                cacheKey = httpContextAccessor.HttpContext.User.GetMsalAccountId();
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/InMemory/MSALPerUserMemoryTokenCacheProvider.cs]: >HttpContext.User. [](start = 47, length = 17)

can any of these values be null?

File: Microsoft.Identity.Web/Client/TokenCacheProviders/Sql/MSALAppSqlTokenCacheProvider.cs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+        /// </summary>
+        /// <param name="args">Contains parameters used by the MSAL call accessing the cache.</param>
+        private void AppTokenCacheBeforeWriteNotification(TokenCacheNotificationArgs args)
+        {
+            // Since we are using a Rowversion for concurrency, we need not to do anything in this handler.
+        }
+
+        /// <summary>
+        /// Raised AFTER MSAL added the new token in its in-memory copy of the cache.
+        /// This notification is called every time MSAL accessed the cache, not just when a write took place:
```
Comment (jennyf19) [inline Microsoft.Identity.Web/Client/TokenCacheProviders/Sql/MSALAppSqlTokenCacheProvider.cs]: >d [](start = 63, length = 1)

accesses

**Comment by jennyf19** [comment]
didn't have time to get through it all...just a few comments.

**Comment by jmprieur** [comment]
@jennyf19 : merging the code (as it fixes issues)
Will get back to the style and comments: cc: @jmprieur

**Pull request merged**